Explain POD and byte array manipulations

km45 opened the pull request:
メモリレイアウトという章を新設して以下の内容を追加します。

* C/C++ における 1バイト
* #40 の POD の説明
* バイト列操作の基礎である memset, memcpy, memcmp の使い方

km45 committed:
Add type-for-byte page

km45 committed:
Explain POD

km45 committed:
Explain std::memset

refer:
http://www.open-std.org/jtc1/sc22/WG14/www/docs/n1256.pdf#%5B%7B%22num%22%3A689%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C-27%2C816%2Cnull%5D

km45 committed:
Polish sentence

km45 committed:
Simplify std::memset usage

km45 committed:
Explain std::memcpy

km45 committed:
Explain std::memcmp

km45 committed:
Add link for std::is_pod

km45 committed:
Add examples for POD

km45 committed:
Rename file

km45 committed:
Move byte array manipulation contents to a new independent page

km45 committed:
Fix header levels following contents movement

km45 committed:
Add comment

km45 committed:
Add class to POD

km45 committed:
Format sentence

km45 committed:
Change section to insert a page

km45 committed:
Add API compatibility for POD page

km45 committed:
Explain that not all C features can be used in C++

km45 committed:
Explain how to change code for C and C++

km45 committed:
Clarify sample code

km45 committed:
Avoid using int for POD in API compatibility page

km45 committed:
Add comment

cpmrph commented on docs/ch12-03-plain-old-data-api-compatibility.md:
```diff
@@ -0,0 +1,178 @@
+
+??? question "C++ だけで名前空間を有効にする"
+    名前空間は名前マングリングに影響しますが、メモリレイアウトには影響しません。
+    名前マングリングが問題になれなければ C++ だけで名前空間を有効にすることができます。
```
なれなければ -> ならなければ

km45 replied:
d32dd0206098874bccce4df138fd2a60510546e2 で修正しました。

cpmrph commented on docs/ch12-02-plain-old-data.md:
```diff
@@ -0,0 +1,207 @@
+- 継承していない
+- メンバ関数を持たない (暗黙的に定義される特別なメンバ関数は明示しない)
+- すべてのメンバ変数は POD
+- メンバ変数に対するアクセス指定子は暗黙的な `public` だけを使用する
```
[imo]
後の 非POD の条件紹介では
`メンバ変数に対するアクセス指定子が2種類以上` と説明しているので、
ここは `メンバ変数に対するアクセス指定子が1種類のみ` でも良いかなと思いました。

km45 replied:
1種類のアクセス指定子を2個以上使用する場合に、
アクセス指定子をまたいでもメモリレイアウトが保証されるか分からなかったために
このような表現にしていました。

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf#chapter.9
> 7 A standard-layout class is a class that:
> — has the same access control (Clause 11) for all non-static data members,

とあり、保証されるようなので提示いただいた表現に修正します。

km45 replied:
6d0b642ffb38773fe1ea3e7c23b638c82336bee1 で変更しました。

暗黙的に `public` 指定されることは明記しておく必要があると思いますので、
引き続き括弧書きで記載しています。

cpmrph commented on docs/ch12-02-plain-old-data.md:
```diff
@@ -0,0 +1,207 @@
+
+#### 非 POD のメンバ変数
+
+`static` ではないメンバ変数に POD ではない型がある構造体は POD ではありません。
```
[imo]
少し読み取りづらく感じました。
`POD ではない型の非 static メンバ変数を持つ構造体は POD ではありません。`
とかはいかがでしょうか。

km45 replied:
1e47b2fbd90caa6777413f0a2fa1d22e788f98a5 で修正しました。

km45 committed:
Fix sentence

km45 committed:
Confirm POD condition and change the description. Discussed at below.
https://github.com/rinatz/cpp-book/pull/41#discussion_r301125159

The struct, which has multiple access specifiers for members and
all of them (including the implicit specifier) are same, is POD.

e.g.) below struct is POD
struct Hoge {
    int x;  // implicit public
 public:
    int y;  // explicit public
};

c.f.) http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf#chapter.9
> 7 A standard-layout class is a class that:
> — has the same access control (Clause 11) for all non-static data members,

km45 committed:
Fix description because don't care static members

km45 committed:
Improve description
